### ISSUE TITLE
Rework regular expression parsing/compiling to be more lenient

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#74]
Duktape's regular expression parser/compiler conforms strictly to the E5/E5.1 regular expression syntax, and rejects many regular expressions that are leniently allowed by other engines. It's confusing to users that other engines parse several regexp forms that Duktape (technically correctly) rejects as a SyntaxError.

Examples of technically invalid regexps that work with most engines:
- `/^\_/`: underscore escape is not allowed, underscore is an IdentifierPart and identity escapes are not allowed for identifier part characters.
- `/\$/`: regexp escape is not allowed, dollar sign is an IdentifierPart. Duktape already has a fix to allow a dollar escape because it's so common.
- `/\\{/`: unescaped brace is not allowed (unless it is part of a quantifier like `/x{5}/`).
- `/[\2]/`: non-zero decimal escape is not allowed (E5 Section 15.10.2.19) but accepted by other engines.

Because these regexps are quite widely used, Duktape should probably parse regexps a bit more leniently. This change is non-trivial because lenient regexp parsing needs backtracking which doesn't fit easily into the current regexp parser/compiler code. For instance, when parsing `/{foo}/`, Duktape will see the left curly brace and think it's parsing a quantifier. When it fails to parse, it currently throws a SyntaxError. Instead, it should rewind and assume `{` is meant literally, i.e. parse the regexp as `/\{foo\}/`.

Related issues with notes on various failing regexps:
- #19
- #39
- #69
- #86
- #142
- #425
- #643

Tasks:
- [ ] Figure out what leniency is necessary for "real world" compatibility with other engines. The main goal is to avoid unnecessarily surprising users.
- [ ] Make the necessary changes to Duktape for more lenient parsing. The changes should be made with code footprint in mind (not growing the footprint if possible). There should be an option to restore compliance, if possible. There's probably no need to have an option to enable/disable individual leniency cases, just a lenient/compliant option.
- [ ] Update testcases.
- [ ] Update internal documentation (`doc/regexp.rst`) to reflect the changes in the regexp algorithm.
- [ ] Update external documentation. Document the non-compliant regexp forms allowed.

**Comment by svaarala** [comment]
By the way, ES6 has optional support for literal braces in Section B.1.4:

- http://www.ecma-international.org/ecma-262/6.0/#sec-regular-expressions-patterns

This doesn't extend to unquoted square brackets etc though.

**Comment by fatcerberus** [comment]
Nice, so we can say the curly brace support is part of "some features borrowed from Ecmascript E6" ;)

**Comment by svaarala** [comment]
Yes, and as a technical detail the config option shouldn't say "DUK_USE_NONSTD_..." but "DUK_USE_ES6_...".

**Comment by mathiasbynens** [comment]
Another example: `/]/` — see https://github.com/slevithan/xregexp/pull/141. (Since portability is one of Duktape’s goals you’ll want to support this.)
